### PR TITLE
Define cirq.Gate**-1 for gates with a _decompose_ method

### PR DESCRIPTION
**Pull request by Strilanc**
- Does nothing for gates with no decomposition or a non-unitary decomposition
- Refactor ReversibleCompositeGate into _InverseCompositeGate
- Test it
- Make Gate validate qubit count

Helps reduce code in https://github.com/quantumlib/Cirq/pull/1244

### CODE REVIEW

**Comment by Strilanc** [comment]
> Why is this showing up so high in the hierarchy.

Because it is a feature that I want to automatically apply to any gate that a user creates that has a `_decompose_` method returning invertable operations.

An alternative I considered was putting it directly into `protocols.inverse`, but this way creates less coupling.

In general, I want a user implementing `Gate` to have useful/obvious tasks "just work". So it makes sense to put such things on `Gate` itself.

> I don't like putting decompose here at all and don't see what this gives us beyond the gates themselves implementing pow or not.

Read through https://github.com/quantumlib/Cirq/pull/1244 . Notice how every gate implemented there has a `self.inverse` attribute, and a `_decompose_` method with a redundant inverse branch? This code fixes that boilerplate.

File: cirq/ops/raw_types.py [hunk 1]
```diff
@@ -95,7 +97,6 @@ class Gate(metaclass=abc.ABCMeta):
     can be used to avoid writing this boilerplate.
     """
 
-    # noinspection PyMethodMayBeStatic
     def validate_args(self, qubits: Sequence[QubitId]) -> None:
         """Checks if this gate can be applied to the given qubits.
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/raw_types.py]: Done.

File: cirq/ops/raw_types.py [hunk 2]
```diff
@@ -107,7 +108,10 @@ def validate_args(self, qubits: Sequence[QubitId]) -> None:
         Throws:
             ValueError: The gate can't be applied to the qubits.
         """
-        pass
+        if len(qubits) != self.num_qubits():
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/raw_types.py]: Done.

**Comment by Strilanc** [comment]
@dabacon Are you still opposed to this change?

**Comment by dabacon** [comment]
Yeah I don't this is correct, I think you will end up regretting having this as the classes become overloaded and the concept of Gate becomes diluted. I also really think decompose is broken and we will see this even more as people encounter magic decompose making their circuits worse.  But won't block you from making this mistake ;)

**Comment by Strilanc** [comment]
Thanks @dabacon .

It's interesting how you see this as diluting Gate. I see it as reinforcing what a gate is, because the class has actual value-add beyond a type check.

**Pull request merged**